### PR TITLE
libuv loop 2.0

### PR DESCRIPTION
**Pull request by ColdenCullen**
Improvements include:
* All subscriptions through the AWS Event Loop API are now edge triggered
* libuv v0.10 and v1.x are both supported
* One of `AWS_USE_IO_COMPLETION_PORTS`, `AWS_USE_EPOLL`, or `AWS_USE_KQUEUE` is always defined.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/libuv/uv_event_loop.c [hunk 1]
```diff
@@ -145,33 +167,94 @@ static uint64_t s_timestamp_to_uv_millis(struct aws_event_loop *event_loop, cons
...
+#if UV_VERSION_MAJOR == 0
+    /* uv won't let us tell it we care about remote hangups, so we have to trick it into thinking it wants it */
+    handle_data->poll.io_watcher.pevents |= POLLRDHUP;
+#else
+    /* Can't use RDHUP trick here, since uv will turn it off in uv__io_stop */
+    handle_data->uv_events |= UV_DISCONNECT;
+#endif
+
+    /* Start listening for events */
+    uv_poll_start(&handle_data->poll, handle_data->uv_events, (uv_poll_cb)0xBAADF00D);
```
Comment (justinboswell) [inline source/libuv/uv_event_loop.c]: Is this value special to libuv, or just obvious? Would be good to make it clearer.

File: source/libuv/uv_event_loop.c [hunk 2]
```diff
@@ -744,17 +834,28 @@ struct aws_event_loop *aws_event_loop_new_libuv(struct aws_allocator *alloc, aws
     struct libuv_loop *impl = NULL;
     struct libuv_owned *owned = NULL;
     uv_loop_t *uv_loop = NULL;
+
+#if UV_VERSION_MAJOR == 0
+    size_t alloc_count = 3;
```
Comment (justinboswell) [inline source/libuv/uv_event_loop.c]: const?

File: source/libuv/uv_event_loop.c [hunk 3]
```diff
@@ -216,7 +216,7 @@ static void s_do_subscribe(struct libuv_loop *impl, struct handle_data *handle_d
     handle_data->uv_events |= UV_DISCONNECT;
 #endif
 
-    /* Start listening for events */
+    /* Start listening for events. The callback should never actually be called, so we use a poison. */
```
Comment (justinboswell) [inline source/libuv/uv_event_loop.c]: Could we not just route it to a callback of our own that asserts fatally? At least then we'd understand the callstack?

File: CMakeLists.txt [hunk 4]
```diff
@@ -77,6 +77,8 @@ if (WIN32)
     set(PLATFORM_LIBS Secur32 Crypt32)
     set(TLS_STACK_DETERMINED ON)
 
+    set(EVENT_LOOP_DEFINE "IO_COMPLETION_PORTS")
```
Comment (graebm) [inline CMakeLists.txt]: You're always setting this whether or not the USE_IO_COMPLETION_PORT options is ON

**Pull request merged**